### ISSUE TITLE
Generalize GitHub URI to allow GitHub Enterprise host

### ISSUE THREAD

**Issue by lgonzalezsa** [fastai/nbdev#916]
Was trying to use `nbdev_release_gh` and was having `fastcore.basics.HTTP401UnauthorizedError: HTTP Error 401: Unauthorized`
Reviewing the code in https://github.com/fastai/nbdev/blob/master/nbs/17_release.ipynb
Specifically the cell:
```
#| export
GH_HOST = "https://api.github.com" 
```

Could be possible to include a parameter en settings.ini with the URI to use?
I have the option to run nbdev in our GitHub Enterprise then the GH_HOST is different.

PD: what could be the best way to reference a specific code line if using Jupyter Notebook? referencing the specific JSON line?

**Comment by seeM** [comment]
@lgonzalezsa Could you please try setting the `GH_HOST` environment variable before running `nbdev_release_gh`?

`nbdev.release` uses `ghapi` which uses that environment variable over here: https://github.com/fastai/ghapi/blob/d86f41815418384ed3cf32a8526c4843d060ce7b/ghapi/core.py#L20

I tend to link to the corresponding .py file when possible. Unfortunately GitHub isn't yet very good at linking to and rendering notebooks. I use the "Copy permalink" feature so that the link always works - rather than linking to a branch with code that's likely to change.

**Comment by lgonzalezsa** [comment]
Based on [GitHub documentation the API schema](https://docs.github.com/en/enterprise-server@3.4/rest/overview/resources-in-the-rest-api#schema) changes a bit and not sure if differs with public REST API.
nbdev defines:
GH_HOST = "https://api.github.com" 

According to GitHub documentation able to test a call with my token, then exported as follow:
GH_HOST = "http(s)://[hostname]/api/v3"
Resulting on:
`pickle.PicklingError: Can't pickle <class 'fastcore.basics.HTTP404NotFoundError'>: attribute lookup HTTP404NotFoundError on fastcore.basics failed`

**Comment by seeM** [comment]
I’m sorry, I don’t follow. Have you tried setting the environment variable I mentioned above? For example:

```sh
GH_HOST=<your enterprise url here> nbdev_release_gh
```

**Comment by seeM** [comment]
Oh, did that cause the pickling error?

**Comment by lgonzalezsa** [comment]
Correct, sorry, was testing exporting the GH_HOST variable following GitHub documentation and that cause the pickling issue.
I did test the endpoint with my token and works, then seems the URI I use is correct, what I am not sure if the API varies from GitHub public to the GitHub Enterprise.

**Comment by seeM** [comment]
I see! Jeremy recently added the ability in ghapi to set the env var `GITHUB_DEBUG=1` to print request info. That might help determine whether the api schema differs

**Comment by lgonzalezsa** [comment]
Testing it but seems same output to me.
![image](https://user-images.githubusercontent.com/704686/186058917-7fa934ea-9487-4e8d-93ee-2ccf483eca15.png)

**Comment by seeM** [comment]
Do you have the latest ghapi installed?

**Comment by lgonzalezsa** [comment]
Updating from 1.0.0 to 1.0.1 same traceback output.
```
(conda_dev) gonluisr@gonluisr-hp-zbook-15-g3:~/Downloads/scripts$ conda list | grep ghapi
ghapi                     1.0.1                      py_0    fastai
```
Testing ghapi directly seems to work:
```bash
ghapi git.get_ref luis-gs linux-setup --ref heads/master
{'ref': 'refs/heads/master', 'node_id': 'MDM6UmVmMTk0MDcxOnJlZnMvaGVhZHMvbWFzdGVy', 'url': 'https://github.enterprise.com/api/v3/repos/luis-gs/linux-setup/git/refs/heads/master', 'object': {'sha': '8d4e3d3fe5b6ab17191eefbdf85664ab1f4377c1', 'type': 'commit', 'url': 'https://github.enterprise.com/api/v3/repos/luis-gs/linux-setup/git/commits/8d4e3d3fe5b6ab17191eefbdf85664ab1f4377c1'}}
```

**Comment by seeM** [comment]
I think this is a minimal example btw. Hard for me to reproduce because I don't have GH enterprise to test with.

```python
from ghapi.core import GhApi
gh = GhApi('luis-gs', 'linux-setup')
gh.issues.list_for_repo(state='closed', sort='created', filter='all', since='2000-01-01T00:00:004Z', labels='enhancement')
```

**Comment by lgonzalezsa** [comment]
I understand, thank you for your help, here the traceback:
```
Unexpected exception formatting exception. Falling back to standard exception
Traceback (most recent call last):
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3398, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "/tmp/ipykernel_281120/3649172761.py", line 1, in <cell line: 1>
    gh.issues.list_for_repo(state='closed', sort='created', filter='all', since='2000-01-01T00:00:004Z', labels='enhancement')
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/ghapi/core.py", line 61, in __call__
    return self.client(self.path, self.verb, headers=headers, route=route_p, query=query_p, data=data_p)
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/ghapi/core.py", line 118, in __call__
    res,self.recv_hdrs = urlsend(path, verb, headers=headers or None, debug=debug, return_headers=True,
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/fastcore/net.py", line 214, in urlsend
    return urlread(req, return_json=return_json, return_headers=return_headers)
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/fastcore/net.py", line 115, in urlread
    if 400 <= e.code < 500: raise ExceptionsHTTP[e.code](e.url, e.hdrs, e.fp) from None
fastcore.basics.HTTP404NotFoundError: HTTP Error 404: Not Found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 1990, in showtraceback
    if hasattr(value, "_render_traceback_"):
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/tempfile.py", line 472, in __getattr__
    file = self.__dict__['file']
KeyError: 'file'
```

**Comment by seeM** [comment]
Okay so it does seem to be a path issue. If you could get an example working for the `issues/list_for_repos` endpoint with enterprise, even if it uses curl, that might be a good start

**Comment by lgonzalezsa** [comment]
the endpoint you mentioned is this one? https://docs.github.com/en/rest/issues/issues#list-repository-issues

**Comment by lgonzalezsa** [comment]
Testing it with curl and works:
```bash
curl   -H "Accept: application/vnd.github+json" -H "Authorization: token :)" https://github.enterprise.com/api/v3/repos/luis-gs/linux-setup/issues
[
  {
    "url": "https://github.enterprise.com/api/v3/repos/luis-gs/linux-setup/issues/3",
    "repository_url": "https://github.enterprise.com/api/v3/repos/luis-gs/linux-setup",
    "labels_url": "https://github.enterprise.com/api/v3/repos/luis-gs/linux-setup/issues/3/labels{/name}",
    "comments_url": "https://github.enterprise.com/api/v3/repos/luis-gs/linux-setup/issues/3/comments",
    "events_url": "https://github.enterprise.com/api/v3/repos/luis-gs/linux-setup/issues/3/events",
    "html_url": "https://github.enterprise.com/luis-gs/linux-setup/issues/3",
    "id": 1770968,
    "node_id": "MDU6SXNzdWUxNzcwOTY4",
    "number": 3,
    "title": "testing issue",
    "user": {
      "login": "luis-gs",
      "id": 19024,
      "node_id": "MDQ6VXNlcjE5MDI0",
      "avatar_url": "https://avatars.github.enterprise.com/u/19024?",
      "gravatar_id": "",
      "url": "https://github.enterprise.com/api/v3/users/luis-gs",
      "html_url": "https://github.enterprise.com/luis-gs",
      "followers_url": "https://github.enterprise.com/api/v3/users/luis-gs/followers",
      "following_url": "https://github.enterprise.com/api/v3/users/luis-gs/following{/other_user}",
      "gists_url": "https://github.enterprise.com/api/v3/users/luis-gs/gists{/gist_id}",
      "starred_url": "https://github.enterprise.com/api/v3/users/luis-gs/starred{/owner}{/repo}",
      "subscriptions_url": "https://github.enterprise.com/api/v3/users/luis-gs/subscriptions",
      "organizations_url": "https://github.enterprise.com/api/v3/users/luis-gs/orgs",
      "repos_url": "https://github.enterprise.com/api/v3/users/luis-gs/repos",
      "events_url": "https://github.enterprise.com/api/v3/users/luis-gs/events{/privacy}",
      "received_events_url": "https://github.enterprise.com/api/v3/users/luis-gs/received_events",
      "type": "User",
      "site_admin": false
    },
    "labels": [

    ],
    "state": "open",
    "locked": false,
    "assignee": null,
    "assignees": [

    ],
    "milestone": null,
    "comments": 0,
    "created_at": "2022-08-24T02:36:26Z",
    "updated_at": "2022-08-24T02:36:26Z",
    "closed_at": null,
    "author_association": "OWNER",
    "active_lock_reason": null,
    "body": null,
    "reactions": {
      "url": "https://github.enterprise.com/api/v3/repos/luis-gs/linux-setup/issues/3/reactions",
      "total_count": 0,
      "+1": 0,
      "-1": 0,
      "laugh": 0,
      "hooray": 0,
      "confused": 0,
      "heart": 0,
      "rocket": 0,
      "eyes": 0
    },
    "timeline_url": "https://github.enterprise.com/api/v3/repos/luis-gs/linux-setup/issues/3/timeline",
    "performed_via_github_app": null
  }
]
```

**Comment by seeM** [comment]
Hmm, I just realised the debug environment variable is actually `GHAPI_DEBUG=1`. Could you try rerunning the ghapi version of above with that? We can see what URL it uses, compare to your curl, and figure out why its 404ing. For example, here's what I get with `GHAPI_DEBUG=1`:

```python
from ghapi.core import GhApi
gh = GhApi('fastai', 'nbdev') # NOTE: replace with your owner + repo
res = gh.issues.list_for_repo(state='closed', sort='created', filter='all', since='2000-01-01T00:00:004Z', labels='enhancement')
```
```python
{'data': None,
 'full_url': 'https://api.github.com/repos/fastai/nbdev/issues?state=closed&labels=enhancement&sort=created&since=2000-01-01T00%3A00%3A004Z',
 'headers': {'Accept': 'application/vnd.github.v3+json'},
 'method': 'GET'}
```

---

It could also be that you need to provide the token via the env var `GITHUB_TOKEN`?

**Comment by lgonzalezsa** [comment]
Seems no difference in API, 
Here result
```bash
In [1]: from ghapi.core import GhApi
   ...: gh = GhApi('luis-gs', 'linux-setup') # NOTE: replace with your owner + repo
   ...: res = gh.issues.list_for_repo(state='closed', sort='created', filter='all', since='2000-01-01T00:00:004Z', labels='enhancement')
{'data': None,
 'full_url': 'https://github.enterprise.com/api/v3/repos/luis-gs/linux-setup/issues?state=closed&labels=enhancement&sort=created&since=2000-01-01T00%3A00%3A004Z',
 'headers': {'Accept': 'application/vnd.github.v3+json'},
 'method': 'GET'}

In [2]: res
Out[2]: (#0) []

In [3]: quit
```

Also running the nbdev_changelog with debug enabled, the endpoints, was testing with curl releases/latest and got "Not Found", not sure if that could be the issue, see below.

```bash
$ nbdev_changelog 
{'data': None,
 'full_url': 'https://github.enterprise.com/api/v3/repos/enterprise/the_repo/releases/latest',
 'headers': {'Accept': 'application/vnd.github.v3+json'},
 'method': 'GET'}
{'data': None,
 'full_url': 'https://github.enterprise.com/api/v3/repos/enterprise/the_repo/issues?state=closed&labels=breaking&sort=created&since=2000-01-01T00%3A00%3A004Z',
 'headers': {'Accept': 'application/vnd.github.v3+json'},
 'method': 'GET'}
{'data': None,
 'full_url': 'https://github.enterprise.com/api/v3/repos/enterprise/the_repo/issues?state=closed&labels=enhancement&sort=created&since=2000-01-01T00%3A00%3A004Z',
 'headers': {'Accept': 'application/vnd.github.v3+json'},
 'method': 'GET'}
{'data': None,
 'full_url': 'https://github.enterprise.com/api/v3/repos/enterprise/the_repo/issues?state=closed&labels=bug&sort=created&since=2000-01-01T00%3A00%3A004Z',
 'headers': {'Accept': 'application/vnd.github.v3+json'},
 'method': 'GET'}
concurrent.futures.process._RemoteTraceback: 
```
Testing with curl
```bash
{
    "message": "Not Found",
    "documentation_url": "https://docs.github.com/enterprise/3.4/rest/reference/repos#get-the-latest-release"
  }
```

**Comment by seeM** [comment]
@lgonzalezsa we’re trying to get access to a GitHub enterprise account to improve its support. Will keep you posted

Could you please share the full stack trace from the most recent nbdev_changelog? Is that with the GITHUB_TOKEN environment variable set?

**Comment by lgonzalezsa** [comment]
Here the complete traceback and yes theGITHUB_TOKEN was set in the environment.
```bash
$ nbdev_changelog 
{'data': None,
 'full_url': 'https://github.enterprise.com/api/v3/repos/enterprise/the_repo/releases/latest',
 'headers': {'Accept': 'application/vnd.github.v3+json'},
 'method': 'GET'}
{'data': None,
 'full_url': 'https://github.enterprise.com/api/v3/repos/enterprise/the_repo/issues?state=closed&labels=breaking&sort=created&since=2000-01-01T00%3A00%3A004Z',
 'headers': {'Accept': 'application/vnd.github.v3+json'},
 'method': 'GET'}
{'data': None,
 'full_url': 'https://github.enterprise.com/api/v3/repos/enterprise/the_repo/issues?state=closed&labels=enhancement&sort=created&since=2000-01-01T00%3A00%3A004Z',
 'headers': {'Accept': 'application/vnd.github.v3+json'},
 'method': 'GET'}
{'data': None,
 'full_url': 'https://github.enterprise.com/api/v3/repos/enterprise/the_repo/issues?state=closed&labels=bug&sort=created&since=2000-01-01T00%3A00%3A004Z',
 'headers': {'Accept': 'application/vnd.github.v3+json'},
 'method': 'GET'}
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/concurrent/futures/process.py", line 243, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/concurrent/futures/process.py", line 202, in _process_chunk
    return [fn(*args) for args in chunk]
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/concurrent/futures/process.py", line 202, in <listcomp>
    return [fn(*args) for args in chunk]
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/fastcore/parallel.py", line 46, in _call
    return g(item)
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/nbdev/release.py", line 62, in _issues
    return self.gh.issues.list_for_repo(state='closed', sort='created', filter='all', since=self.commit_date, labels=label)
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/ghapi/core.py", line 61, in __call__
    return self.client(self.path, self.verb, headers=headers, route=route_p, query=query_p, data=data_p)
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/ghapi/core.py", line 118, in __call__
    res,self.recv_hdrs = urlsend(path, verb, headers=headers or None, debug=debug, return_headers=True,
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/fastcore/net.py", line 214, in urlsend
    return urlread(req, return_json=return_json, return_headers=return_headers)
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/fastcore/net.py", line 115, in urlread
    if 400 <= e.code < 500: raise ExceptionsHTTP[e.code](e.url, e.hdrs, e.fp) from None
fastcore.basics.HTTP404NotFoundError: HTTP Error 404: Not Found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/concurrent/futures/process.py", line 208, in _sendback_result
    result_queue.put(_ResultItem(work_id, result=result,
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/multiprocessing/queues.py", line 372, in put
    obj = _ForkingPickler.dumps(obj)
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <class 'fastcore.basics.HTTP404NotFoundError'>: attribute lookup HTTP404NotFoundError on fastcore.basics failed
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/gonluisr/miniconda3/envs/conda_dev/bin/nbdev_changelog", line 10, in <module>
    sys.exit(changelog())
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/fastcore/script.py", line 117, in _f
    return tfunc(**merge(args, args_from_prog(func, xtra)))
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/nbdev/release.py", line 108, in changelog
    res = Release(repo=repo).changelog(debug=debug)
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/nbdev/release.py", line 75, in changelog
    issues = self._issue_groups()
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/nbdev/release.py", line 63, in _issue_groups
    def _issue_groups(self): return parallel(self._issues, self.groups.keys(), progress=False)
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/fastcore/parallel.py", line 117, in parallel
    return L(r)
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/fastcore/foundation.py", line 98, in __call__
    return super().__call__(x, *args, **kwargs)
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/fastcore/foundation.py", line 106, in __init__
    items = listify(items, *rest, use_list=use_list, match=match)
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/site-packages/fastcore/basics.py", line 66, in listify
    elif is_iter(o): res = list(o)
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/concurrent/futures/process.py", line 559, in _chain_from_iterable_of_lists
    for element in iterable:
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/concurrent/futures/_base.py", line 609, in result_iterator
    yield fs.pop().result()
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/concurrent/futures/_base.py", line 446, in result
    return self.__get_result()
  File "/home/gonluisr/miniconda3/envs/conda_dev/lib/python3.9/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
_pickle.PicklingError: Can't pickle <class 'fastcore.basics.HTTP404NotFoundError'>: attribute lookup HTTP404NotFoundError on fastcore.basics failed
```

**Comment by lgonzalezsa** [comment]
@seeM today I did another try with nbdev_new using GH_HOST=https://github.enterprise.com/api/v3 and GHAPI_DEBUG=1
Everything goes well until (from my understanding) the cloning of nbdev_template which certainly does not exist in my GitHub Enterprise: 'full_url': 'https://github.hpe.com/api/v3/repos/fastai/nbdev-template/releases/latest'

Seems to me is needed to separate the GH_HOST for my repos from the GH_HOST of the template.

CC: @hamelsmu 

```bash
{'data': b'{"description": "Python package for data analysis in '
         b'n", "homepage": "https://enterprise.github.io/the_repo'
         b's_libs"}',
 'full_url': 'https://github.enterprise.com/api/v3/repos/enterprise/the_repo',
 'headers': {'Accept': 'application/vnd.github.v3+json'},
 'method': 'PATCH'}
{'data': None,
 'full_url': 'https://github.enterprise.com/api/v3/repos/fastai/nbdev-template/releases/latest',
 'headers': {'Accept': 'application/vnd.github.v3+json'},
 'method': 'GET'}
Traceback (most recent call last):
  File "/home/data_analysis/.conda/envs/gtcs-dataanalysis/bin/nbdev_new", line 8, in <module>
    sys.exit(nbdev_new())
  File "/home/data_analysis/.conda/envs/gtcs-dataanalysis/lib/python3.9/site-packages/fastcore/script.py", line 119, in _f
    return tfunc(**merge(args, args_from_prog(func, xtra)))
  File "/home/data_analysis/.conda/envs/gtcs-dataanalysis/lib/python3.9/site-packages/nbdev/cli.py", line 88, in nbdev_new
    tag = GhApi().repos.get_latest_release('fastai', 'nbdev-template').tag_name
  File "/home/data_analysis/.conda/envs/gtcs-dataanalysis/lib/python3.9/site-packages/ghapi/core.py", line 61, in __call__
    return self.client(self.path, self.verb, headers=headers, route=route_p, query=query_p, data=data_p)
  File "/home/data_analysis/.conda/envs/gtcs-dataanalysis/lib/python3.9/site-packages/ghapi/core.py", line 118, in __call__
    res,self.recv_hdrs = urlsend(path, verb, headers=headers or None, debug=debug, return_headers=True,
  File "/home/data_analysis/.conda/envs/gtcs-dataanalysis/lib/python3.9/site-packages/fastcore/net.py", line 214, in urlsend
    return urlread(req, return_json=return_json, return_headers=return_headers)
  File "/home/data_analysis/.conda/envs/gtcs-dataanalysis/lib/python3.9/site-packages/fastcore/net.py", line 115, in urlread
    if 400 <= e.code < 500: raise ExceptionsHTTP[e.code](e.url, e.hdrs, e.fp) from None
fastcore.basics.HTTP404NotFoundError: HTTP Error 404: Not Found
```

**Comment by seeM** [comment]
@lgonzalezsa we’ve fixed that in latest master. Slow progress, but progress nonetheless 😄 could you please try again?

**Comment by seeM** [comment]
Btw @lgonzalezsa I’d be open to pairing on this issue over a call if you like. We really want to get nbdev working well with GH Enterprise, and that might be the fastest way until we figure out a better testing environment

**Comment by lgonzalezsa** [comment]
Was testing the latest in master:
```bash~/nbdev$ git log
commit c1f1b4f4d5c3de9e423921a560c724eb8c6c7ed0 (HEAD -> master, origin/master, origin/HEAD)
Merge: 08462b3 92ec4ea
Author: Wasim Lorgat <mwlorgat@gmail.com>
Date:   Thu Sep 8 10:42:51 2022 +1000

    Merge pull request #1009 from seeM/nbdev-template-gh-host
```
Results:
```bash
nbdev_new 
{'data': None,
 'full_url': 'https://github.enterprise.com/api/v3/repos/enterprise/the_repo',
 'headers': {'Accept': 'application/vnd.github.v3+json'},
 'method': 'GET'}
repo = the_repo # Automatically inferred from git
user = enterprise # Automatically inferred from git
author = Luis Rolando Gonzalez # Automatically inferred from git
author_email = luis.gs@enterprise.com # Automatically inferred from git
description = To support FPY Process Improvements # Automatically inferred from git
settings.ini created.
{'data': b'{"description": "To support Process Improvements", "homepage": "'
         b'https://enterprise.github.io/the_repo"}',
 'full_url': 'https://github.enterprise.com/api/v3/repos/enterprise/the_repo',
 'headers': {'Accept': 'application/vnd.github.v3+json'},
 'method': 'PATCH'}
{'data': None,
 'full_url': 'https://api.github.com/repos/fastai/nbdev-template/releases/latest',
 'headers': {'Accept': 'application/vnd.github.v3+json'},
 'method': 'GET'}
Traceback (most recent call last):
  File "/home/data_analysis/.conda/envs/module/bin/nbdev_new", line 8, in <module>
    sys.exit(nbdev_new())
  File "/home/data_analysis/.conda/envs/module/lib/python3.9/site-packages/fastcore/script.py", line 119, in _f
    return tfunc(**merge(args, args_from_prog(func, xtra)))
  File "/home/data_analysis/.conda/envs/module/lib/python3.9/site-packages/nbdev/cli.py", line 88, in nbdev_new
    tag = GhApi(gh_host='https://api.github.com').repos.get_latest_release('fastai', 'nbdev-template').tag_name
  File "/home/data_analysis/.conda/envs/module/lib/python3.9/site-packages/ghapi/core.py", line 61, in __call__
    return self.client(self.path, self.verb, headers=headers, route=route_p, query=query_p, data=data_p)
  File "/home/data_analysis/.conda/envs/module/lib/python3.9/site-packages/ghapi/core.py", line 118, in __call__
    res,self.recv_hdrs = urlsend(path, verb, headers=headers or None, debug=debug, return_headers=True,
  File "/home/data_analysis/.conda/envs/module/lib/python3.9/site-packages/fastcore/net.py", line 217, in urlsend
    return urlread(req, return_json=return_json, return_headers=return_headers)
  File "/home/data_analysis/.conda/envs/module/lib/python3.9/site-packages/fastcore/net.py", line 118, in urlread
    if 400 <= e.code < 500: raise ExceptionsHTTP[e.code](e.url, e.hdrs, e.fp, msg=e.msg) from None
fastcore.basics.HTTP401UnauthorizedError: HTTP Error 401: Unauthorized
====Error Body====
{
  "message": "Bad credentials",
  "documentation_url": "https://docs.github.com/rest"
}
```
Seems that now the code is pointing to the right GH_HOST for the nbdev-template but trying to use my GITHUB_TOKEN for the GitHub Enterprise to authenticate? 
 'full_url': 'https://api.github.com/repos/fastai/nbdev-template/releases/latest'

I am willing to connect in a call as well! that will be awesome, let me know how I can collaborate.

**Comment by seeM** [comment]
@lgonzalezsa thanks for the call! :) I appreciate you making the time, and glad we figured out the `nbdev_changelog` issues. I just realised that you might have had a `token` file in the root of you repo that was stopping nbdev from using the `GITHUB_TOKEN` environment variable

**Comment by seeM** [comment]
Closing. Both `nbdev.release` (#1024) and `nbdev_new` (#1007) should now be working on GitHub Enterprise. `nbdev.release` also respects the `GITHUB_TOKEN` environment variable (in addition to `NBDEV_TOKEN` and a `token` file, #1025). @lgonzalezsa please create another issue if you encounter any problems :)

The error you saw with `nbdev_new` is because your token was invalid. IIUC, any `GhApi` call fails if you have an invalid token set.

**Comment by lgonzalezsa** [comment]
Thanks @seeM good session indeed. With regards the token, then the issue was the name of the environment variable right? I need to use NBDEV_TOKEN instead of GITHUB_TOKEN, think that's how works at the end. Now, because nbdev_new is cloning the template from github.com certainly my GitHub token seems is not working because is for the Enterprise not for GitHub.com. Or what you are saying is that I need to use the two environmental variables, one for GH Enterprise and another for GitHub.com?
Just cloned master and tested:
```bash
nbdev                     2.3.0                    pypi_0    pypi
```
traceback including extra print statements:

```bash
nbdev_new
repo = the_repo # Automatically inferred from git
user = enterprise # Automatically inferred from git
author = Luis Rolando Gonzalez # Automatically inferred from git
author_email = luis.gs@enterprise.com # Automatically inferred from git
description = To support FPY Process Improvements # Automatically inferred from git
/home/gonluisr/miniconda3/envs/dev_env/lib/python3.10/site-packages/nbdev/config.py:171: UserWarning: Config file already exists: /home/gonluisr/github_enterprise/the_repo/settings.ini and will be used as a base
  if cfg.config_file.exists(): warn(f'Config file already exists: {cfg.config_file} and will be used as a base')
settings.ini created.
401
https://api.github.com/repos/fastai/nbdev-template/releases/latest
Server: GitHub.com
Date: Sat, 10 Sep 2022 14:58:31 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 90
X-GitHub-Media-Type: github.v3; format=json
X-RateLimit-Limit: 60
X-RateLimit-Remaining: 58
X-RateLimit-Reset: 1662822888
X-RateLimit-Used: 2
X-RateLimit-Resource: core
Access-Control-Expose-Headers: ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset
Access-Control-Allow-Origin: *
Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
X-Frame-Options: deny
X-Content-Type-Options: nosniff
X-XSS-Protection: 0
Referrer-Policy: origin-when-cross-origin, strict-origin-when-cross-origin
Content-Security-Policy: default-src 'none'
Vary: Accept-Encoding, Accept, X-Requested-With
X-GitHub-Request-Id: 4DC2:57AF:3CA5DEA:7C0F166:631CA617
connection: close


<http.client.HTTPResponse object at 0x7f3461133370>
Unauthorized
====Error Body====
{
  "message": "Bad credentials",
  "documentation_url": "https://docs.github.com/rest"
}

Traceback (most recent call last):
  File "/home/gonluisr/miniconda3/envs/dev_env/bin/nbdev_new", line 8, in <module>
    sys.exit(nbdev_new())
  File "/home/gonluisr/miniconda3/envs/dev_env/lib/python3.10/site-packages/fastcore/script.py", line 119, in _f
    return tfunc(**merge(args, args_from_prog(func, xtra)))
  File "/home/gonluisr/miniconda3/envs/dev_env/lib/python3.10/site-packages/nbdev/cli.py", line 88, in nbdev_new
    tag = GhApi(gh_host='https://api.github.com').repos.get_latest_release('fastai', 'nbdev-template').tag_name
  File "/home/gonluisr/miniconda3/envs/dev_env/lib/python3.10/site-packages/ghapi/core.py", line 61, in __call__
    return self.client(self.path, self.verb, headers=headers, route=route_p, query=query_p, data=data_p)
  File "/home/gonluisr/miniconda3/envs/dev_env/lib/python3.10/site-packages/ghapi/core.py", line 118, in __call__
    res,self.recv_hdrs = urlsend(path, verb, headers=headers or None, debug=debug, return_headers=True,
  File "/home/gonluisr/miniconda3/envs/dev_env/lib/python3.10/site-packages/fastcore/net.py", line 222, in urlsend
    return urlread(req, return_json=return_json, return_headers=return_headers)
  File "/home/gonluisr/miniconda3/envs/dev_env/lib/python3.10/site-packages/fastcore/net.py", line 123, in urlread
    if 400 <= e.code < 500: raise ExceptionsHTTP[e.code](e.url, e.hdrs, e.fp, msg=e.msg) from None

```

**Comment by seeM** [comment]
@lgonzalezsa with latest master, can now use `GITHUB_TOKEN` instead of `NBDEV_TOKEN`. My understanding is that the 401 is because whichever token it’s using is not a valid GitHub token. For example if I do this I also get a 401 because `foo` is not a valid token:

```sh
GITHUB_TOKEN=foo nbdev_new
``

I think you might have one of the environment variables set to an invalid token, or you might have an old `token` file in your repo.

You can test if your token is valid by doing a GET https://api.github.com/octocat request. Or something like `ghapi repos.get fastai nbdev-template`.

**Comment by lgonzalezsa** [comment]
Sounds good, now only using GITHUB_TOKEN, but as mentioned before when using GitHub Enterprise repos, nbdev_new seems communicates to both GitHub hosts, to the enterprise related to my repo and to GitHub.com to clone nbdev-template, as shown in the last traceback.
No issues with `https://github.enterprise.com/api/v3/repos/enterprise/the_repo` but when trying to reach nbdev-template I got 401.
```bash
401
https://api.github.com/repos/fastai/nbdev-template/releases/latest
Server: GitHub.com
```
The token I created in GitHub Enterprise does not work with GitHub.com, exclusively works with the enterprise repos.
I tested with curl also and the token works fine with GHE, or maybe am doing something wrong ¯\_(ツ)_/¯

**Comment by seeM** [comment]
@lgonzalezsa ohhh that makes perfect sense. I understand now. I’ll make a separate issue and tag you

**Comment by seeM** [comment]
@lgonzalezsa the issue is being tracked in:
- https://github.com/fastai/nbdev/issues/1038

**Comment by seeM** [comment]
@lgonzalezsa #1038 is merged. Could you please try `nbdev_new` again with the latest nbdev **and ghapi**?

Have you had any other issues with GitHub Enterprise Server? If not, I think we can consider it supported and close https://github.com/fastai/nbdev/issues/944

**Comment by lgonzalezsa** [comment]
Sorry for delay! Awesome `nbdev_new` works with no issues now.
```bash
nbdev_new
repo = the_repo # Automatically inferred from git
user = enterprise # Automatically inferred from git
author = Luis Rolando Gonzalez # Automatically inferred from git
author_email = luis.gs@enterprise.com # Automatically inferred from git
description = Dask get_file demo # Automatically inferred from git
/home/gonluisr/miniconda3/envs/dev_env/lib/python3.10/site-packages/nbdev/config.py:173: UserWarning: Config file already exists: /home/gonluisr/github_enterprise/the_repo/settings.ini and will be used as a base
  if cfg.config_file.exists(): warn(f'Config file already exists: {cfg.config_file} and will be used as a base')
settings.ini created.
pandoc -o README.md
  to: gfm+footnotes+tex_math_dollars-yaml_metadata_block
  output-file: index.html
  standalone: true
  default-image-extension: png
  
metadata
  description: Dask get_file demo
  title: the_repo
  
Output created: _docs/README.md

```

**Comment by seeM** [comment]
Great @lgonzalezsa! I've closed #944 too. Please let us know if you encounter any problems :)
